Interactivity API: Add `wp-each` directive (#57859)
WordPress/gutenberg: packages/interactivity/docs/2-api-reference.md

Reviewer: `wp-each` has not been added to the Table of Contents

    - [`wp-each`](#wp-each) ![](https://img.shields.io/badge/TEMPLATING-afd2e3.svg)

Author: Nice catch!

#### `wp-each`

The `wp-each` directive is intended to render a list of elements. The directive can be used in `<template>` tags, being the value a path to an array stored in the global state or the context. The content inside the `<template>` tag is the template used to render each of the items.

Each item is included in the context under the `item` name by default, so directives inside the template can access the current item.

For example, let's consider the following HTML.

```html
<ul data-wp-context='{ "list": [ "hello", "hola", "olá" ] }'>
  <template data-wp-each="context.list" >
    <li data-wp-text="context.item"></li>
  </template>
</ul>
```

It would generate the following output:

```html
<ul data-wp-context='{ "list": [ "hello", "hola", "olá" ] }'>
  <li data-wp-text="context.item">hello</li>
  <li data-wp-text="context.item">hola</li>
  <li data-wp-text="context.item">olá</li>
</ul>
```

The prop that holds the item in the context can be changed by passing a suffix to the directive name. In the following example, we change the default prop `item` to `greeting`.

```html
<ul data-wp-context='{ "list": [ "hello", "hola", "olá" ] }'>

>   <template data-wp-each--greeting="context.list" >

Reviewer: What is the use case to use this solution instead of the previous one?
` <template data-wp-each="context.list" >
    <li data-wp-text="context.item"></li>`
, make it easier to read?

PD: Just asking for learn! (more than a proper review). I did not see anything weird.

Reviewer 2: Mostly for nested lists.

```js
store("myPlugin", {
  state: {
    names: ["Jon", "Aegon"],
    surnames: ["Snow", "Targaryen"],
    get fullname() {
      const { name, surname } = getContext();
      return `${name} ${surname}`;
    },
  },
});
```

```js
<ul>
  <template data-wp-each--name="state.names">
    <template data-wp-each--surname="state.surnames">
      <li data-wp-text="state.fullname"></li>
    </template>
  </template>
</ul>
```

This will render:

```js
<ul>
  <li>Jon Snow</li>
  <li>Jon Targaryen</li>
  <li>Aegon Snow</li>
  <li>Aegon Targaryen</li>
</ul>
```

Reviewer: Cool. We should add this one or a similar example in the docs. But beware of spoilers! 😄

Reviewer 2: > But beware of spoilers! 😄

Ups 😆

    <li data-wp-text="context.greeting"></li>
  </template>
</ul>
```

By default, it uses each element as the key for the rendered nodes, but you can also specify a path to retrieve the key if necessary, e.g., when the list contains objects.

For that, you must use `data-wp-each-key` in the `<template>` tag and not `data-wp-key` inside the template content. This is because `data-wp-each` creates
a context provider wrapper around each rendered item, and those wrappers are the ones that need the `key` property.

```html
<ul data-wp-context='{
  "list": [
    { "id": "en", "value": "hello" },
    { "id": "es", "value": "hola" },
    { "id": "pt", "value": "olá" }
  ]
}'>
  <template
    data-wp-each--greeting="context.list"
    data-wp-each-key="context.greeting.id"
  >
    <li data-wp-text="context.greeting.value"></li>
  </template>
</ul>
```

> For server-side rendered lists, another directive called `data-wp-each-child` ensures hydration works as expected. It should be placed in the server-side rendered elements.

Reviewer 2: Let's make sure people don't think that they need to take care of this:

```suggestion
For server-side rendered lists, another directive called `data-wp-each-child` ensures hydration works as expected. This directive is added automatically when the directive is processed on the server.
```
